File: fractal/src/main/java/net/fwitz/math/fractal/escape/DefaultEscapeFunction.java
package net.fwitz.math.fractal.escape;

import net.fwitz.math.binary.complex.Complex;

import java.util.OptionalDouble;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;

import static java.util.Objects.requireNonNull;

class DefaultEscapeFunction implements EscapeFunction {
    private final Function<Complex, Complex> init;
    private final Predicate<Complex> shortcutContainmentTest;
    private final Predicate<Complex> escapeTest;
    private final BiFunction<Complex, Complex, Complex> step;
    private final int maxIters;
    private final boolean includeInit;
    private final OptionalDouble convergenceAdjust;
    private final OptionalDouble divergenceAdjust;

    DefaultEscapeFunction(DefaultBuilder builder) {
        this.init = builder.init;
        this.shortcutContainmentTest = builder.shortcutContainmentTest;
        this.escapeTest = requireNonNull(builder.escapeTest, "escapeTest or containmentTest");
        this.step = requireNonNull(builder.step, "step");
        this.maxIters = requireNonNull(builder.maxIters, "maxIters");
        this.includeInit = builder.includeInit;
        this.convergenceAdjust = builder.convergenceSmoothingFactor.isPresent()
                ? OptionalDouble.of(builder.convergenceSmoothingFactor.getAsDouble() / maxIters)
                : OptionalDouble.empty();
        this.divergenceAdjust = builder.divergenceSmoothingFactor.isPresent()
                ? OptionalDouble.of(builder.divergenceSmoothingFactor.getAsDouble() / maxIters)
                : OptionalDouble.empty();
    }

    public EscapeTimeResult apply(Complex c) {
        Complex zOld = Complex.ZERO;

        // Note that if a shortcut test is used, then the fractal generation is faster, but since no
        // iterations are performed, the final value is the initial value rather than whatever iteration
        // would have produced and no convergence information is available.
        if (shortcutContainmentTest.test(c)) {
            return EscapeTimeResult.contained(c, OptionalDouble.empty());
        }

        int i = 0;
        Complex z = init.apply(c);
        Smoother cSmooth = convergenceAdjust.isPresent()
                ? new ConvergenceSmoother(convergenceAdjust.getAsDouble(), includeInit ? z : c)
                : Smoother.NULL;
        Smoother dSmooth = divergenceAdjust.isPresent()
                ? new DivergenceSmoother(divergenceAdjust.getAsDouble())
                : Smoother.NULL;

        if (includeInit) {
            ++i;
            cSmooth.accept(z);
            dSmooth.accept(z);
        }

        if (escapeTest.test(z)) {
            return escaped(i, z, dSmooth);
        }

        while (++i <= maxIters) {
            z = step.apply(c, z);
            cSmooth.accept(z);
            dSmooth.accept(z);
            if (escapeTest.test(z)) {
                return escaped(i, z, dSmooth);
            }
        }

        return contained(z, cSmooth);
    }

    private EscapeTimeResult escaped(int iters, Complex z, Smoother dSmooth) {
        return EscapeTimeResult.escaped(iters, z, dSmooth.finish());
    }

    private EscapeTimeResult contained(Complex z, Smoother cSmooth) {
        return EscapeTimeResult.contained(z, cSmooth.finish());
    }

    static Builder builder() {
        return new DefaultBuilder();
    }


    private interface Smoother extends Consumer<Complex> {
        Smoother NULL = z -> {
        };

        default OptionalDouble finish() {
            return OptionalDouble.empty();
        }
    }

    private static class ConvergenceSmoother implements Smoother {
        private final double adjust;

        private Complex zOld;
        private double sum;

        private ConvergenceSmoother(double adjust, Complex zOld) {
            this.adjust = adjust;
            this.zOld = zOld;
        }

        @Override
        public void accept(Complex z) {
            sum += adjust * Math.exp(-1 / zOld.minus(z).abs());
            zOld = z;
        }

        @Override
        public OptionalDouble finish() {
            return OptionalDouble.of(sum);
        }
    }

    private static class DivergenceSmoother implements Smoother {
        private final double adjust;

        private double sum;

        private DivergenceSmoother(double adjust) {
            this.adjust = adjust;
        }

        @Override
        public void accept(Complex z) {
            sum += adjust * Math.exp(-1 / z.abs());
        }

        @Override
        public OptionalDouble finish() {
            return OptionalDouble.of(sum);
        }
    }


    private static class DefaultBuilder implements Builder {
        Function<Complex, Complex> init = c -> c;
        Predicate<Complex> shortcutContainmentTest = c -> false;
        boolean includeInit;
        OptionalDouble convergenceSmoothingFactor = OptionalDouble.empty();
        OptionalDouble divergenceSmoothingFactor = OptionalDouble.empty();

        Predicate<Complex> escapeTest;
        BiFunction<Complex, Complex, Complex> step;
        Integer maxIters;

        private DefaultBuilder() {
        }

        int maxIters() {
            return maxIters;
        }

        public Builder init(Function<Complex, Complex> init) {
            this.init = requireNonNull(init);
            return this;
        }

        @Override
        public Builder shortcutContainmentTest(Predicate<Complex> shortcutContainmentTest) {
            this.shortcutContainmentTest = requireNonNull(shortcutContainmentTest);
            return this;
        }

        public Builder containmentTest(Predicate<Complex> containmentTest) {
            this.escapeTest = containmentTest.negate();
            return this;
        }

        public Builder escapeTest(Predicate<Complex> escapeTest) {
            this.escapeTest = requireNonNull(escapeTest);
            return this;
        }

        public Builder step(BiFunction<Complex, Complex, Complex> step) {
            this.step = requireNonNull(step);
            return this;
        }

        public Builder maxIters(int maxIters) {
            if (maxIters < 1) {
                throw new IllegalArgumentException("maxIters < 1: " + maxIters);
            }
            this.maxIters = maxIters;
            return this;
        }

        public Builder includeInit() {
            this.includeInit = true;
            return this;
        }

        @Override
        public Builder enableConvergenceSmoothing(double adjustFactor) {
            convergenceSmoothingFactor = OptionalDouble.of(adjustFactor);
            return this;
        }

        @Override
        public Builder enableDivergenceSmoothing(double adjustFactor) {
            divergenceSmoothingFactor = OptionalDouble.of(adjustFactor);
            return this;
        }

        public EscapeFunction build() {
            return new DefaultEscapeFunction(this);
        }
    }
}

File: fractal/src/main/kotlin/net/fwitz/math/fractal/escape/DefaultEscapeFunction.kt
package net.fwitz.math.fractal.escape

import net.fwitz.math.binary.BinaryNumber
import kotlin.math.exp

internal class DefaultEscapeFunction<T : BinaryNumber<T>> private constructor(
    builder: DefaultBuilder<T>
) : EscapeFunction<T> {
    private val init: (T) -> T
    private val shortcutContainmentTest: (T) -> Boolean
    private val escapeTest: (T) -> Boolean
    private val step: (T, T) -> T
    private val maxIters: Int
    private val includeInit: Boolean
    private val convergenceAdjust: Double?
    private val divergenceAdjust: Double?

    companion object {
        fun <T : BinaryNumber<T>> builder(): EscapeFunction.Builder<T> {
            return DefaultBuilder()
        }
    }

    override fun invoke(c: T): EscapeTimeResult<T> {
        // Note that if a shortcut test is used, then the fractal generation is faster, but since no
        // iterations are performed, the final value is the initial value rather than whatever iteration
        // would have produced and no convergence information is available.
        if (shortcutContainmentTest(c)) return EscapeTimeResult.contained(c, null)

        var i = 0
        var z = init(c)
        val cSmooth: Smoother<T> =
            if (convergenceAdjust == null) NullSmoother()
            else ConvergenceSmoother(convergenceAdjust, if (includeInit) z else c)
        val dSmooth: Smoother<T> =
            if (divergenceAdjust == null) NullSmoother()
            else DivergenceSmoother(divergenceAdjust)

        if (includeInit) {
            ++i
            cSmooth(z)
            dSmooth(z)
        }
        if (escapeTest(z)) return escaped(i, z, dSmooth)

        while (++i <= maxIters) {
            z = step(c, z)
            cSmooth(z)
            dSmooth(z)
            if (escapeTest(z)) return escaped(i, z, dSmooth)
        }

        return contained(z, cSmooth)
    }

    private fun escaped(iters: Int, z: T, dSmooth: Smoother<T>): EscapeTimeResult<T> {
        return EscapeTimeResult.escaped(iters, z, dSmooth.finish())
    }

    private fun contained(z: T, cSmooth: Smoother<T>): EscapeTimeResult<T> {
        return EscapeTimeResult.contained(z, cSmooth.finish())
    }

    interface Smoother<T : BinaryNumber<T>> : (T) -> Unit {
        fun finish(): Double? = null
    }

    private class NullSmoother<T : BinaryNumber<T>> : Smoother<T> {
        override fun invoke(z: T) = Unit
    }

    private class ConvergenceSmoother<T : BinaryNumber<T>>(
        val adjust: Double,
        private var zOld: T
    ) : Smoother<T> {
        private var sum = 0.0

        override fun invoke(z: T) {
            sum += adjust * exp(-1 / (zOld - z).abs)
            zOld = z
        }

        override fun finish() = sum
    }

    private class DivergenceSmoother<T : BinaryNumber<T>>(
        val adjust: Double
    ) : Smoother<T> {
        private var sum = 0.0

        override fun invoke(z: T) {
            sum += adjust * exp(-1 / z.abs)
        }

        override fun finish() = sum
    }

    private class DefaultBuilder<T : BinaryNumber<T>> : EscapeFunction.Builder<T> {
        var init: (T) -> T = { it }
        var shortcutContainmentTest: (T) -> Boolean = { false }
        var includeInit = false
        var convergenceSmoothingFactor: Double? = null
        var divergenceSmoothingFactor: Double? = null
        var escapeTest: ((T) -> Boolean)? = null
        var step: ((T, T) -> T)? = null
        var maxIters: Int? = null

        fun maxIters() = maxIters!!

        override fun init(init: (T) -> T) = this.also {
            this.init = init
        }

        override fun shortcutContainmentTest(shortcutContainmentTest: (T) -> Boolean) = this.also {
            this.shortcutContainmentTest = shortcutContainmentTest
        }

        override fun containmentTest(containmentTest: (T) -> Boolean) = this.also {
            escapeTest = { !containmentTest(it) }
        }

        override fun escapeTest(escapeTest: (T) -> Boolean) = this.also {
            this.escapeTest = escapeTest
        }

        override fun step(step: (T, T) -> T) = this.also {
            this.step = step
        }

        override fun maxIters(maxIters: Int) = this.also {
            require(maxIters >= 1) { "maxIters < 1: $maxIters" }
            this.maxIters = maxIters
        }

        override fun includeInit() = this.also {
            includeInit = true
        }

        override fun enableConvergenceSmoothing(adjustFactor: Double) = this.also {
            convergenceSmoothingFactor = adjustFactor
            return this
        }

        override fun enableDivergenceSmoothing(adjustFactor: Double) = this.also {
            divergenceSmoothingFactor = adjustFactor
            return this
        }

        override fun build(): EscapeFunction<T> {
            return DefaultEscapeFunction(this)
        }
    }

    init {
        init = builder.init
        shortcutContainmentTest = builder.shortcutContainmentTest
        escapeTest = requireNotNull(builder.escapeTest) { "escapeTest or containmentTest" }
        step = requireNotNull(builder.step) { "step" }
        maxIters = requireNotNull(builder.maxIters) { "maxIters" }
        includeInit = builder.includeInit
        convergenceAdjust = builder.convergenceSmoothingFactor?.let { it / maxIters }
        divergenceAdjust = builder.divergenceSmoothingFactor?.let { it / maxIters }
    }
}